### PR TITLE
Add caching for required charsets

### PR DESCRIPTION
**Pull request by LinuxClient**
I have seen the following comment and thought to do this:
`// maybe cache?`

### CODE REVIEW

**Comment by LinuxClient** [comment]
I excuse me if this PR is a bit large. I've omitted to create a separate PR.

**Comment by LinuxClient** [comment]
I don't know if I keep overlooking things, but I can't find the errors that Travis CI says.

File: src/main/java/org/apache/commons/io/Charsets.java [hunk 1]
```diff
@@ -59,6 +59,17 @@
     // correctly and without delay on all Java platforms.
     //
 
+    private static final TreeMap<String, Charset> REQUIRED_CHARSETS = new TreeMap<>(String.CASE_INSENSITIVE_ORDER);
```
Comment (michael-o) [inline src/main/java/org/apache/commons/io/Charsets.java:62]: Use the interface name

File: src/main/java/org/apache/commons/io/Charsets.java [hunk 2]
```diff
@@ -72,15 +83,7 @@
      * @since 2.5
      */
     public static SortedMap<String, Charset> requiredCharsets() {
```
Comment (michael-o) [inline src/main/java/org/apache/commons/io/Charsets.java:85]: Why sorted map? TreeMap is sorted already?!

**Comment by garydgregory** [comment]
There are too many unrelated changes here as you are changing 19 filles. Please focus the PR on what you really want to do.

File: src/main/java/org/apache/commons/io/Charsets.java [hunk 3]
```diff
@@ -71,16 +85,8 @@
...
-        // maybe cache?
-        final TreeMap<String, Charset> m = new TreeMap<>(String.CASE_INSENSITIVE_ORDER);
-        m.put(StandardCharsets.ISO_8859_1.name(), StandardCharsets.ISO_8859_1);
-        m.put(StandardCharsets.US_ASCII.name(), StandardCharsets.US_ASCII);
-        m.put(StandardCharsets.UTF_16.name(), StandardCharsets.UTF_16);
-        m.put(StandardCharsets.UTF_16BE.name(), StandardCharsets.UTF_16BE);
-        m.put(StandardCharsets.UTF_16LE.name(), StandardCharsets.UTF_16LE);
-        m.put(StandardCharsets.UTF_8.name(), StandardCharsets.UTF_8);
-        return Collections.unmodifiableSortedMap(m);
+    public static Map<String, Charset> requiredCharsets() {
```
[the remark this answers is not in the document]
Comment (LinuxClient) [reply src/main/java/org/apache/commons/io/Charsets.java]: Yeah, I saw it on the Travis CI Log. But it was mentioned by @michael-o above.

**Pull request merged**